### PR TITLE
Use multihosts feature from client-go

### PR DESCRIPTION
**Pull request by dshulyak**
Next options are available for a client that wants to tolerate api server failures:
- virtual ip, eg managed by keepalived, with all clients connected to that ip
- dns a records, and rely on golang client to try all of them
- use kubernetes service ip
First two options are providing addiotnal overhead, especially for environmentsthat are not running in the cloud infrastructure.And in some cases it is simply not possible to rely on kubernetes service ip.

This change allows to use multiple server endpoints in RESTClient and rotate them if necessary. During normal operation RESTClient will stick to a single, randomly selected, endpoint. And new one will be selected only If requests to the api server will start to fail repeatedly.

kubectl config set-cluster was extended with additional options to configure multiple servers:

```
kubectl config set-cluster dind --servers=http://10.0.0.1,http://10.0.0.2

clusters:
- cluster:
    servers:
    - http://10.0.0.1
    - http://10.0.0.2
  name: test
```

And --server and --servers are not mutually exclusive:

```
kubectl config set-cluster dind --servers=http://10.0.0.2 --server=http://10.0.0.1

clusters:
- cluster:
    server: http://10.0.0.1
    servers:
    - http://10.0.0.1
```

**Release note:**
```
Multiple api-server endpoints can be configured by kubectl config set-cluster and used by RESTClient.
```

### CODE REVIEW

File: pkg/kubectl/cmd/config/create_cluster.go [hunk 1]
```diff
@@ -124,6 +126,9 @@ func (o *createClusterOptions) modifyCluster(existingCluster clientcmdapi.Cluste
 	if o.server.Provided() {
 		modifiedCluster.Server = o.server.Value()
 	}
+	if len(o.servers) > 0 {
```
Comment (brendandburns) [inline pkg/kubectl/cmd/config/create_cluster.go]: Should we check that a user doesn't specify both `--server` and `--servers`?

File: staging/src/k8s.io/client-go/rest/config.go [hunk 2]
```diff
@@ -304,12 +324,24 @@ func InClusterConfig() (*Config, error) {
 //
 // Note: the Insecure flag is ignored when testing for this value, so MITM attacks are
 // still possible.
-func IsConfigTransportTLS(config Config) bool {
-	baseURL, _, err := defaultServerUrlFor(&config)
+func IsConfigTransportTLS(config Config) (bool, error) {
+	hosts, _, err := defaultServerUrlsFor(&config)
+	allHttps := 0
```
Comment (brendandburns) [inline staging/src/k8s.io/client-go/rest/config.go]: move this down to where it is first used?

**Review by brendandburns — COMMENTED** [review]
Some minor comments.

**Comment by mattymo** [comment]
This PR is quite useful for any application that requires a kubeconfig (kubectl, kubefed, kubelet)

**Comment by dshulyak** [comment]
@k8s-bot gce etcd3 e2e test thi

File: staging/src/k8s.io/client-go/rest/config.go (hunk 2 above)
Comment (dshulyak) [inline staging/src/k8s.io/client-go/rest/config.go]: fixed

File: pkg/kubectl/cmd/config/create_cluster.go (hunk 1 above)
Comment (dshulyak) [inline pkg/kubectl/cmd/config/create_cluster.go]: I remember that during initial discussion we agreed that both values will be used if provided, so if user will specify both server and servers we will create array [server, servers...] without duplications

**Comment by dshulyak** [comment]
Sorry it took me too long to get back to this patch, but may i ask for review please? @brendandburns @krousey

File: staging/src/k8s.io/client-go/rest/config.go [hunk 3]
```diff
@@ -52,6 +52,7 @@ type Config struct {
 	// be appended to all request URIs used to access the apiserver. This allows a frontend
 	// proxy to easily relocate all of the apiserver endpoints.
 	Host string
+
```
Comment (brendandburns) [inline staging/src/k8s.io/client-go/rest/config.go]: nit: delete added whitespace.

File: staging/src/k8s.io/client-go/rest/urlcontainer.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	"sync"
+	"time"
+
+	"k8s.io/client-go/util/flowcontrol"
+)
+
+// NewURLContainer initialezes URLContainer instance and returns pointer,
+// randomly selects URL that will be considered valid
+func NewURLContainer(urls []*url.URL) *URLContainer {
+	var errorps float32 = 5
```
Comment (brendandburns) [inline staging/src/k8s.io/client-go/rest/urlcontainer.go]: nit: making these variables seems pointless, just inline the constants.

File: staging/src/k8s.io/client-go/rest/urlcontainer.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+		return
+	}
+	if !c.ratelimiter.TryAccept() {
+		c.renewStickyURL()
+		c.renewRateLimiter()
+	}
+}
+
+func findRngIntExcludingProvided(rng *rand.Rand, toExclude, length int) int {
+	for {
```
Comment (brendandburns) [inline staging/src/k8s.io/client-go/rest/urlcontainer.go]: this makes me uncomfortable, test your pre-conditions and panic if length == 0, or 1 and `toExclude` == 0?

**Comment by brendandburns** [comment]
Generally LGTM, some nits.

Needs a rebase...

**Comment by brendandburns** [comment]
/retest

@dshulyak please add a meaningful description of this PR so that release note automation will have details.

Once the tests are clean, I think we can LGTM.

**Comment by vadapalliravikumar** [comment]
kube-proxy is not mentioned in this PR. do we need a separate mechanism for the kube-proxy to use multiple apiservers ? or is that covered by this PR ?

And in which release can we expect this. Is it possible to back port this to 1.6

**Comment by kfox1111** [comment]
Any update on this?

**Comment by dshulyak** [comment]
@brendanburns please check it out when you will have time

**Comment by bogdando** [comment]
@dshulyak thanks for keeping working on this. The change is really needed for "native app-sided HA"!

**Comment by deitch** [comment]
@brendandburns you approved, but looks like it is going to auto-close. Can we get it merged in?

@dshulyak I think you will need to rebase

**Comment by brendandburns** [comment]
@deitch I can't merge it without the rebase.

If you are interested, feel free to take this PR, do the rebase and then send your own PR with the original commits as well as the rebase.

Thanks!

**Comment by deitch** [comment]
@brendandburns I will give it a day or two to see if @dshulyak takes care of it. If not, will open new PR.

**Comment by weiwei04** [comment]
Ping @dshulyak

@deitch If @dshulyak gives you push perimission, I think rebase on the current pr based on
https://help.github.com/articles/committing-changes-to-a-pull-request-branch-created-from-a-fork/ is a better choice instead of open a new one, since this pr has lgtmed :)

Looking forward to see this feature to be merged, thanks :)

**Comment by mikedanese** [comment]
@kubernetes/sig-api-machinery-pr-reviews

**Comment by mikedanese** [comment]
This also has some api changes in clientcmd api group @kubernetes/sig-api-machinery-api-reviews

**Comment by tsaridas** [comment]
ping @dshulyak

**Comment by hzxuzhonghu** [comment]
Is there anyone working on this?  @dshulyak @deitch  If you have no bandwidth, I can help rebase and open a new one?

**Comment by cmluciano** [comment]
ping @dshulyak ^

**Comment by krmayankk** [comment]
@dshulyak are you still working on this , i can work on this if no one has picked it up

**Comment by liggitt** [comment]
Closing for lack of activity. If there is interest in pursuing this, this seems like a fundamental enough change to warrant a KEP.

/close

**Comment by agile6v** [comment]
/reopen

We really need this feature. Hope it could continue. 
Thanks.

**Comment by agile6v** [comment]
> /reopen
> 
> @hzxuzhonghu
> We really need this feature. Hope it could continue.
> Thanks.

Hello @hzxuzhonghu @dshulyak @brendandburns 

How to reopen it ?
Thanks.

**Comment by hzxuzhonghu** [comment]
I think only the author and k8s collaborator can reopen.

**Pull request closed without merging**